curvefs/src/metaserver/s3compact_wq_impl.cpp

```diff
@@ -0,0 +1,415 @@
+                it->compaction = compaction;
+                it->chunkoff = chunkoff;
+                it->zero = zero;
+                if (end > nodeEnd) {
+                    addToValidList(nodeEnd, end, chunkid, compaction, chunkoff,
+                                   zero);
+                }
+                return;
+            } else if (begin >= nodeBegin && end <= nodeEnd) {
+                it->begin = end;
+                auto front = validList.emplace(it, begin, end, chunkid,
+                                               compaction, chunkoff, zero);
+                if (nodeBegin < begin) {
+                    validList.emplace(front, nodeBegin, begin, nodeChunkid,
+                                      nodeCompaction, nodeChunkoff, nodeZero);
+                }
+                return;
+            } else if (end > nodeEnd && begin > nodeBegin && begin < nodeEnd) {
+                it->begin = begin;
+                it->chunkid = chunkid;
+                it->compaction = compaction;
+                it->chunkoff = chunkoff;
+                it->zero = zero;
+                validList.emplace(it, nodeBegin, begin, nodeChunkid,
+                                  nodeCompaction, nodeChunkoff, nodeZero);
+                addToValidList(nodeEnd, end, chunkid, compaction, chunkoff,
+                               zero);
+                return;
+            } else if (begin < nodeBegin && end < nodeEnd && end > nodeBegin) {
+                it->begin = end;
+                validList.emplace(it, begin, end, chunkid, compaction, chunkoff,
+                                  zero);
+                return;
+            }
+        }
+        validList.emplace_back(begin, end, chunkid, compaction, chunkoff, zero);
+    };
+
+    for (auto i = 0; i < s3chunkinfolist.s3chunks_size(); i++) {
+        auto chunkinfo = s3chunkinfolist.s3chunks(i);
+        addToValidList(chunkinfo.offset(), chunkinfo.offset() + chunkinfo.len(),
+                       chunkinfo.chunkid(), chunkinfo.compaction(),
+                       chunkinfo.offset(), chunkinfo.zero());
+    }
+
+    // merge same chunkid continuous nodes
+    for (auto curr = validList.begin();;) {
+        auto next = std::next(curr);
+        if (next == validList.end()) break;
+        if (curr->end == next->begin) {
+            if (curr->chunkid == next->chunkid) {
+                next->begin = curr->begin;
+                validList.erase(curr);
+            }
+        }
+        curr = next;
+    }
+    return validList;
+}
+
+int S3CompactWorkQueueImpl::ReadFullChunk(
+    const std::list<struct S3CompactWorkQueueImpl::Node> validList,
+    uint64_t inodeId, uint64_t blockSize, std::string* fullChunk,
+    uint64_t* newChunkid, uint64_t* newCompaction) {
+    for (auto curr = validList.begin(); curr != validList.end();) {
+        auto next = std::next(curr);
+        if (curr->zero) {
+            fullChunk->append(curr->end - curr->begin, '\0');
```

Comment:
I think that if node is zero or incontinuity，no need  to merge

Reply:
we only focus on chunk level data, not chunkinfo level